# CoNLL_2.ipynb

This notebook contains the first part of the model training and analysis code from our CoNLL-2020 paper, ["Identifying Incorrect Labels in the CoNLL-2003 Corpus"](https://www.aclweb.org/anthology/2020.conll-1.16/).

If you're new to the Text Extensions for Pandas library, we recommend that you start
by reading through the notebook [`Analyze_Model_Outputs.ipynb`](https://github.com/CODAIT/text-extensions-for-pandas/blob/master/notebooks/Analyze_Model_Outputs.ipynb), which explains the 
portions of the library that we use in the notebooks in this directory.

### Summary

This notebook aggregates the outputs of the 16 models submitted to the CoNLL-2003 competition. Then it compares these outputs to the corpus's gold-standard labels and identifies areas where there is a strong consensus between model outputs coupled with a disagreement with the corpus labels. Finally, it writes out CSV files containing ranked lists of potentially-incorrect labels.

In [1]:
# INITIALIZATION BOILERPLATE

# Libraries
import os
import sys
import numpy as np
import pandas as pd
from typing import *

# And of course we need the text_extensions_for_pandas library itself.
try:
    import text_extensions_for_pandas as tp
except ModuleNotFoundError as e:
    raise Exception("text_extensions_for_pandas package not found on the Jupyter "
                    "kernel's path. Please either run:\n"
                    "   ln -s ../../text_extensions_for_pandas .\n"
                    "from the directory containing this notebook, or use a Python "
                    "environment on which you have used `pip` to install the package.")

# Code shared among notebooks is kept in tp.cleaning, in this directory.
import text_extensions_for_pandas.cleaning
#constants
PROJECT_ROOT ='../..'


In [2]:
# Download and cache the data set.
# NOTE: This data set is licensed for research use only. Be sure to adhere
#  to the terms of the license when using this data set!
data_set_info = tp.io.conll.maybe_download_conll_data("outputs")
data_set_info

{'train': 'outputs/eng.train',
 'dev': 'outputs/eng.testa',
 'test': 'outputs/eng.testb'}

In [3]:
# Load up the same gold standard data we used in Part 1.
gold_standard = tp.io.conll.conll_2003_to_dataframes(
    data_set_info["test"], ["pos", "phrase", "ent"], [False, True, True])
gold_standard = [
    df.drop(columns=["pos", "phrase_iob", "phrase_type"])
    for df in gold_standard
]

# Dictionary from (collection, offset within collection) to dataframe
gold_standard_spans = {("test", i): 
                       tp.io.conll.iob_to_spans(gold_standard[i]) 
                       for i in range(len(gold_standard))}

In [4]:
# Load up the results from all 16 teams at once.
teams = ["bender", "carrerasa", "carrerasb", "chieu", "curran",
         "demeulder", "florian", "hammerton", "hendrickx",
         "klein", "mayfield", "mccallum", "munro", "whitelaw",
         "wu", "zhang"]

# Read all the output files into one dataframe per <document, team> pair.
outputs = { 
    t: tp.io.conll.conll_2003_output_to_dataframes(
        gold_standard, f"{PROJECT_ROOT}/resources/conll_03/ner/results/{t}/eng.testb")
    for t in teams
}  # Type: Dict[str, List[pd.DataFrame]]

# As an example of what we just loaded, show the token metadata for the 
# "mayfield" team's model's output on document 3.
outputs["mayfield"][3]

span ent_iob ent_type  \
0       [0, 10): '-DOCSTART-'       O     None   
1       [11, 20): 'FREESTYLE'       O     None   
2    [21, 33): 'SKIING-WORLD'       O     None   
3             [34, 37): 'CUP'       B     MISC   
4           [38, 43): 'MOGUL'       O     None   
..                        ...     ...      ...   
161      [803, 809): 'Allais'       I      PER   
162           [810, 811): '('       O     None   
163      [811, 817): 'France'       B      LOC   
164           [817, 818): ')'       O     None   
165       [819, 824): '21.58'       O     None   

                                              sentence  
0                                [0, 10): '-DOCSTART-'  
1    [11, 52): 'FREESTYLE SKIING-WORLD CUP MOGUL RE...  
2    [11, 52): 'FREESTYLE SKIING-WORLD CUP MOGUL RE...  
3    [11, 52): 'FREESTYLE SKIING-WORLD CUP MOGUL RE...  
4    [11, 52): 'FREESTYLE SKIING-WORLD CUP MOGUL RE...  
..                                                 ...  
161    [791, 824): '10. Katleen Allais (France) 21.58'  
162    [791, 824): '10. Katleen Allais (France) 21.58'  
163    [791, 824): '10. Katleen Allais (France) 21.58'  
164    [791, 824): '10. Katleen Allais (France) 21.58'  
165    [791, 824): '10. Katleen Allais (France) 21.58'  

[166 rows x 4 columns]

In [5]:
# Convert results from IOB2 tags to spans across all teams and documents
# See https://en.wikipedia.org/wiki/Inside%E2%80%93outside%E2%80%93beginning_(tagging) for details on IOB2 format.
output_spans = {
    t: {("test", i): tp.io.conll.iob_to_spans(outputs[t][i]) 
        for i in range(len(outputs[t]))}
    for t in teams
}  # Type: Dict[str, Dict[Tuple[str, int], pd.DataFrame]]

# As an example, show the first 10 spans that the "florian" team's model
# found on document 2.
output_spans["florian"][("test", 2)].head(10)

span ent_type
0                 [35, 40): 'JAPAN'      LOC
1                 [50, 55): 'SYRIA'      LOC
2                [57, 63): 'AL-AIN'      LOC
3  [65, 85): 'United Arab Emirates'      LOC
4               [144, 149): 'Japan'      LOC
5           [169, 178): 'Asian Cup'     MISC
6               [192, 197): 'Syria'      LOC
7       [209, 222): 'Takuya Takagi'      PER
8         [297, 308): 'Salem Bitar'      PER
9              [403, 409): 'Syrian'     MISC

In [6]:
# Use Pandas merge to find what spans match up exactly for each team's
# results.
# Unlike Part 1, we perform the join across all entity types, looking for
# matches of both the extracted span *and* the entity type label.
#
# Text Extensions for Pandas includes a utility function 
# compute_accuracy_by_document() to makes this collection-level computation
# simpler.
stats = {
    t: tp.io.conll.compute_accuracy_by_document(
    gold_standard_spans, output_spans[t]) 
    for t in teams
}

# Show the result quality statistics by document for the "carrerasa" team
stats["carrerasa"]

fold  doc_num  num_true_positives  num_extracted  num_entities  \
0    test        0                  42             48            45   
1    test        1                  43             44            44   
2    test        2                  51             52            54   
3    test        3                  43             44            44   
4    test        4                  14             19            19   
..    ...      ...                 ...            ...           ...   
226  test      226                   7              7             7   
227  test      227                  19             22            21   
228  test      228                  22             27            27   
229  test      229                  26             27            27   
230  test      230                  22             27            28   

     precision    recall        F1  
0     0.875000  0.933333  0.903226  
1     0.977273  0.977273  0.977273  
2     0.980769  0.944444  0.962264  
3     0.977273  0.977273  0.977273  
4     0.736842  0.736842  0.736842  
..         ...       ...       ...  
226   1.000000  1.000000  1.000000  
227   0.863636  0.904762  0.883721  
228   0.814815  0.814815  0.814815  
229   0.962963  0.962963  0.962963  
230   0.814815  0.785714  0.800000  

[231 rows x 8 columns]

In [7]:
# F1 for document 4 is looking a bit low. Is that just fluke, or is it
# part of a larger trend? 
# In Part 1, we showed how to drill down to and examine "problem" documents.
# Since we have all this additional data, let's try a broader, more 
# quantitative approach. We'll start by building up some more fine-grained 
# data about congruence between the gold standard and the model outputs.
# Pandas' outer join will tell us what entities showed up just in the gold
# standard, just in the model output, or in both sets.
# For starters, let's do this just for the "carrerasa" team and document 4.
doc_id = ("test", 4)
(gold_standard_spans[doc_id]
    .merge(output_spans["carrerasa"][doc_id], how="outer", indicator=True)
    .sort_values("span"))

span ent_type      _merge
0              [19, 28): 'ASIAN CUP'     MISC        both
1                 [46, 52): 'AL-AIN'      LOC        both
2   [54, 74): 'United Arab Emirates'      LOC        both
3             [97, 106): 'Asian Cup'     MISC        both
4                [141, 146): 'Japan'      LOC   left_only
5                [141, 146): 'Japan'      ORG  right_only
6                [149, 154): 'Syria'      LOC   left_only
7                [149, 154): 'Syria'      ORG  right_only
8                [181, 186): 'Japan'      LOC        both
9         [188, 200): 'Hassan Abbas'      PER        both
10       [214, 227): 'Takuya Takagi'      PER        both
11               [232, 237): 'Syria'      LOC        both
12      [239, 253): 'Nader Jokhadar'      PER        both
13               [276, 281): 'China'      LOC   left_only
14               [276, 281): 'China'      ORG  right_only
15          [284, 294): 'Uzbekistan'      LOC   left_only
16          [284, 294): 'Uzbekistan'      ORG  right_only
17       [321, 334): 'Shkvyrin Igor'      PER        both
18      [339, 353): 'Shatskikh Oleg'      PER        both
19          [462, 472): 'Uzbekistan'      LOC        both
20               [487, 492): 'Japan'      LOC   left_only
21               [487, 492): 'Japan'      ORG  right_only
22               [507, 512): 'Syria'      LOC        both
23               [527, 532): 'China'      LOC        both

In [8]:
# Repeat the analysis from the previous cell across all teams and documents.
# That is, perform an outer join between the gold standard spans dataframe
# for each document and the corresponding dataframe from each team.
def merge_span_sets(team: str,
                    gold_results: Dict[Tuple[str, int], pd.DataFrame],
                    results_by_team: Dict[str, Dict[Tuple[str, int], pd.DataFrame]]):
    result = {}  # Type: Dict[Tuple[str, int]: pd.DataFrame]
    for k in gold_results.keys():
        merged = gold_results[k].merge(results_by_team[team][k],
                                       how="outer", indicator=True)
        merged["gold"] = merged["_merge"].isin(("both", "left_only"))
        merged[team] = merged["_merge"].isin(("both", "right_only"))
        result[k] = merged[["span", "ent_type", "gold", team]]
    return result

span_flags = {t: merge_span_sets(t, gold_standard_spans, output_spans) for t in teams}  # Type: Dict[Tuple[str, int]: pd.DataFrame]

In [9]:
# Now we have indicator variables for every extracted span, telling whether 
# it was in the gold standard data set and/or in each of the team's results.
# For example, here are the first 5 spans for document 2 in the "carrerasa"
# team's results:
doc_id = ("test", 2)
span_flags["carrerasa"][doc_id].head(5)

span ent_type  gold  carrerasa
0                 [35, 40): 'JAPAN'      LOC  True       True
1                 [50, 55): 'SYRIA'      LOC  True       True
2                [57, 63): 'AL-AIN'      LOC  True       True
3  [65, 85): 'United Arab Emirates'      LOC  True       True
4               [144, 149): 'Japan'      LOC  True       True

In [10]:
# Do an n-way merge of all those indicator variables across documents.
# This operation produces a single summary dataframe per document.
indicators = {}  # Type: Dict[Tuple[str, int], pd.DataFrame]
for k in gold_standard_spans.keys():
    result = gold_standard_spans[k]
    for t in teams:
        result = result.merge(span_flags[t][k], how="outer")
    #indicators[k] = result.fillna(False)
    indicators[k] = result.infer_objects(copy=False)
    
# Now we have a vector of indicator variables for every span extracted 
# from every document across all the model outputs and the gold standard.
# For example, let's show the results for document 10:
doc_10 = ("test", 10)
indicators[doc_10]

span ent_type   gold bender carrerasa carrerasb  \
0          [11, 22): 'RUGBY UNION'      ORG   True   True      True      True   
1               [24, 30): 'LITTLE'      LOC  False    NaN       NaN       NaN   
2               [24, 30): 'LITTLE'      PER   True   True      True      True   
3         [34, 46): 'MISS CAMPESE'      LOC  False    NaN       NaN       NaN   
4         [34, 46): 'MISS CAMPESE'     MISC  False    NaN       NaN       NaN   
..                             ...      ...    ...    ...       ...       ...   
149  [2397, 2410): 'Andrew Blades'      PER   True   True      True      True   
150         [2397, 2403): 'Andrew'      PER  False    NaN       NaN       NaN   
151   [2415, 2427): 'Marco Caputo'      PER   True   True      True      True   
152          [2415, 2420): 'Marco'      PER  False    NaN       NaN       NaN   
153    [2432, 2443): 'Dan Crowley'      PER   True   True      True      True   

    chieu curran demeulder florian hammerton hendrickx klein mayfield  \
0    True   True      True    True      True      True  True     True   
1     NaN   True       NaN     NaN      True      True   NaN     True   
2    True  False     False   False     False     False  True    False   
3     NaN   True       NaN     NaN       NaN       NaN   NaN      NaN   
4     NaN    NaN       NaN     NaN       NaN       NaN   NaN     True   
..    ...    ...       ...     ...       ...       ...   ...      ...   
149  True   True      True    True     False      True  True     True   
150   NaN    NaN       NaN     NaN      True       NaN   NaN      NaN   
151  True   True      True    True     False      True  True     True   
152   NaN    NaN       NaN     NaN      True       NaN   NaN      NaN   
153  True   True      True    True      True      True  True     True   

    mccallum  munro whitelaw     wu zhang  
0       True   True     True   True  True  
1        NaN   True      NaN    NaN   NaN  
2      False  False    False  False  True  
3        NaN    NaN      NaN    NaN   NaN  
4        NaN    NaN     True    NaN   NaN  
..       ...    ...      ...    ...   ...  
149     True   True     True   True  True  
150      NaN    NaN      NaN    NaN   NaN  
151     True   True     True   True  True  
152      NaN    NaN      NaN    NaN   NaN  
153     True   True     True   True  True  

[154 rows x 19 columns]

In [11]:
# If you look at the above dataframe, you can see that some entities 
# ("RUGBY UNION", for example) are "easy", in that almost every entry
# found them correctly. Other entities, like "CAMPESE", are "harder",
# in that few of the entrants correctly identified them. Let's add
# a column that quantifies this "difficulty level" by counting how 
# many teams found each true or false positive.
for df in indicators.values():
    # Convert the teams' indicator columns into a single matrix of 
    # Boolean values, and sum the number of True values in each row.
    vectors = df[df.columns[3:]].values
    counts = np.count_nonzero(vectors, axis=1)
    df["num_teams"] = counts

# Show the dataframe for document 10 again, this time with the new
# "num_teams" column at the far right.
indicators[doc_10]

span ent_type   gold bender carrerasa carrerasb  \
0          [11, 22): 'RUGBY UNION'      ORG   True   True      True      True   
1               [24, 30): 'LITTLE'      LOC  False    NaN       NaN       NaN   
2               [24, 30): 'LITTLE'      PER   True   True      True      True   
3         [34, 46): 'MISS CAMPESE'      LOC  False    NaN       NaN       NaN   
4         [34, 46): 'MISS CAMPESE'     MISC  False    NaN       NaN       NaN   
..                             ...      ...    ...    ...       ...       ...   
149  [2397, 2410): 'Andrew Blades'      PER   True   True      True      True   
150         [2397, 2403): 'Andrew'      PER  False    NaN       NaN       NaN   
151   [2415, 2427): 'Marco Caputo'      PER   True   True      True      True   
152          [2415, 2420): 'Marco'      PER  False    NaN       NaN       NaN   
153    [2432, 2443): 'Dan Crowley'      PER   True   True      True      True   

    chieu curran demeulder florian hammerton hendrickx klein mayfield  \
0    True   True      True    True      True      True  True     True   
1     NaN   True       NaN     NaN      True      True   NaN     True   
2    True  False     False   False     False     False  True    False   
3     NaN   True       NaN     NaN       NaN       NaN   NaN      NaN   
4     NaN    NaN       NaN     NaN       NaN       NaN   NaN     True   
..    ...    ...       ...     ...       ...       ...   ...      ...   
149  True   True      True    True     False      True  True     True   
150   NaN    NaN       NaN     NaN      True       NaN   NaN      NaN   
151  True   True      True    True     False      True  True     True   
152   NaN    NaN       NaN     NaN      True       NaN   NaN      NaN   
153  True   True      True    True      True      True  True     True   

    mccallum  munro whitelaw     wu zhang  num_teams  
0       True   True     True   True  True         16  
1        NaN   True      NaN    NaN   NaN         16  
2      False  False    False  False  True          6  
3        NaN    NaN      NaN    NaN   NaN         16  
4        NaN    NaN     True    NaN   NaN         16  
..       ...    ...      ...    ...   ...        ...  
149     True   True     True   True  True         15  
150      NaN    NaN      NaN    NaN   NaN         16  
151     True   True     True   True  True         15  
152      NaN    NaN      NaN    NaN   NaN         16  
153     True   True     True   True  True         16  

[154 rows x 20 columns]

In [12]:
# Now we can rank the entities in document 10 by "difficulty", either as 
# true positives for the models to find...
# (just for document 10 for the moment)
ind = indicators[doc_10].copy()
ind[ind["gold"] == True].sort_values("num_teams").head(10)

span ent_type  gold bender carrerasa carrerasb  \
7          [39, 46): 'CAMPESE'      PER  True  False     False     False   
45  [1018, 1028): 'Barbarians'      ORG  True  False     False     False   
75   [1687, 1696): 'All Black'      ORG  True  False     False     False   
23  [333, 345): 'Queenslander'     MISC  True  False     False     False   
66  [1535, 1545): 'Barbarians'      ORG  True  False      True     False   
15    [163, 173): 'Barbarians'      ORG  True   True     False     False   
54  [1332, 1342): 'Twickenham'      LOC  True   True     False     False   
2           [24, 30): 'LITTLE'      PER  True   True      True      True   
84  [1740, 1750): 'Barbarians'      ORG  True  False      True     False   
40     [759, 768): 'Wallabies'      ORG  True  False      True     False   

    chieu curran demeulder florian hammerton hendrickx  klein mayfield  \
7    True  False     False    True     False     False  False    False   
45  False  False     False   False     False     False   True    False   
75  False  False     False   False     False     False  False    False   
23   True  False      True    True     False     False  False    False   
66  False  False     False   False     False      True   True    False   
15  False  False      True   False     False      True  False    False   
54  False   True     False   False     False      True   True     True   
2    True  False     False   False     False     False   True    False   
84  False  False     False    True     False     False  False     True   
40   True   True     False    True     False      True  False    False   

   mccallum  munro whitelaw     wu  zhang  num_teams  
7     False  False    False  False  False          2  
45     True  False    False  False  False          2  
75    False   True     True   True  False          3  
23    False  False    False  False   True          4  
66     True   True    False  False  False          5  
15     True   True    False  False  False          5  
54    False  False    False  False  False          5  
2     False  False    False  False   True          6  
84     True   True     True  False  False          6  
40     True   True    False   True  False          8

In [13]:
# ...or as false positives to avoid:
ind[ind["gold"] == False].sort_values("num_teams", ascending=False).head(10)

span ent_type   gold bender carrerasa carrerasb  \
1            [24, 30): 'LITTLE'      LOC  False    NaN       NaN       NaN   
109    [2003, 2011): 'McIntosh'      PER  False    NaN       NaN       NaN   
73    [1687, 1696): 'All Black'      LOC  False    NaN       NaN       NaN   
74    [1687, 1696): 'All Black'     MISC  False    NaN       NaN       NaN   
76    [1687, 1696): 'All Black'      PER  False    NaN      True       NaN   
77        [1691, 1696): 'Black'      LOC  False   True       NaN      True   
78        [1691, 1696): 'Black'      PER  False    NaN       NaN       NaN   
81         [1720, 1724): 'Norm'      PER  False    NaN       NaN       NaN   
82   [1740, 1750): 'Barbarians'      LOC  False    NaN       NaN       NaN   
83   [1740, 1750): 'Barbarians'     MISC  False    NaN       NaN       NaN   

    chieu curran demeulder florian hammerton hendrickx klein mayfield  \
1     NaN   True       NaN     NaN      True      True   NaN     True   
109   NaN    NaN       NaN     NaN       NaN      True   NaN      NaN   
73    NaN    NaN       NaN     NaN       NaN       NaN  True      NaN   
74   True   True       NaN    True       NaN       NaN   NaN      NaN   
76    NaN    NaN       NaN     NaN       NaN       NaN   NaN      NaN   
77    NaN    NaN       NaN     NaN       NaN       NaN   NaN      NaN   
78    NaN    NaN      True     NaN       NaN      True   NaN     True   
81    NaN    NaN       NaN     NaN      True       NaN   NaN      NaN   
82    NaN    NaN      True     NaN       NaN       NaN   NaN      NaN   
83   True   True       NaN     NaN       NaN       NaN  True      NaN   

    mccallum munro whitelaw   wu zhang  num_teams  
1        NaN  True      NaN  NaN   NaN         16  
109      NaN   NaN      NaN  NaN   NaN         16  
73      True   NaN      NaN  NaN   NaN         16  
74       NaN   NaN      NaN  NaN  True         16  
76       NaN   NaN      NaN  NaN   NaN         16  
77       NaN   NaN      NaN  NaN   NaN         16  
78       NaN   NaN      NaN  NaN   NaN         16  
81       NaN   NaN      NaN  NaN   NaN         16  
82       NaN   NaN      NaN  NaN   NaN         16  
83       NaN   NaN      NaN  NaN  True         16

In [14]:
# To get a better picture of what entities are "difficult", we need to look 
# across the entire test set. Let's combine the dataframes in 
# `indicators` into a single dataframe that covers all the documents.

# First we preprocess each dataframe to make it easier to combine.
to_stack = [
    pd.DataFrame({
        "fold": k[0],  # Keys are (collection, offset) tuples
        "doc_offset": k[1],
        "span" : indicators[k]["span"],
        "ent_type": indicators[k]["ent_type"],
        "gold": indicators[k]["gold"],
        "num_teams": indicators[k]["num_teams"]
    })
    for k in indicators.keys()
    #for i in range(len(indicators))
]

# Then we concatenate all the preprocessed dataframes into a single dataframe.
counts = pd.concat(to_stack)
counts

fold  doc_offset                               span ent_type   gold  \
0   test           0                  [19, 24): 'JAPAN'      LOC   True   
1   test           0                  [19, 24): 'JAPAN'     MISC  False   
2   test           0                  [29, 34): 'LUCKY'     MISC  False   
3   test           0                  [29, 34): 'LUCKY'      PER  False   
4   test           0                    [35, 38): 'WIN'      PER  False   
..   ...         ...                                ...      ...    ...   
50  test         230     [1341, 1355): '1966 World Cup'     MISC   True   
51  test         230  [1346, 1363): 'World Cup winning'     MISC  False   
52  test         230          [1346, 1355): 'World Cup'     MISC  False   
53  test         230              [1395, 1400): 'Bobby'      LOC  False   
54  test         230              [1395, 1400): 'Bobby'      PER   True   

    num_teams  
0          12  
1          16  
2          16  
3          16  
4          16  
..        ...  
50          1  
51         16  
52         16  
53         16  
54         14  

[11999 rows x 6 columns]

In [15]:
# Now we can pull out the most difficult entities across the entire test
# set.
# First, let's find the most difficult entities from the standpoint of recall:
# entities that are in the gold standard, but not in most results.
difficult_recall = counts[counts["gold"] == True].sort_values("num_teams").reset_index(drop=True)
difficult_recall.head(10)

fold  doc_offset                                   span ent_type  gold  \
0  test         163                [587, 596): 'Newsnight'     MISC  True   
1  test         152                     [1487, 1489): 'Po'      LOC  True   
2  test         170                [461, 470): 'Myos Yang'     MISC  True   
3  test         176                 [57, 67): 'K.T. Arasu'      LOC  True   
4  test         199         [108, 124): 'Scottish premier'     MISC  True   
5  test         199  [27, 52): 'SCOTTISH PREMIER DIVISION'     MISC  True   
6  test          97             [141, 153): 'Bahia Blanca'      LOC  True   
7  test          97                  [133, 140): 'Rosario'      LOC  True   
8  test          97                  [125, 132): 'Quequen'      LOC  True   
9  test          63             [148, 160): 'Conservative'     MISC  True   

   num_teams  
0          0  
1          0  
2          0  
3          0  
4          0  
5          0  
6          0  
7          0  
8          0  
9          0

In [16]:
# Hmm, everything is zero. How many entities were found by zero teams?  One team?
(counts[counts["gold"] == True][["num_teams", "span"]]
 .groupby("num_teams").count()
 .rename(columns={"span": "count"}))

count
num_teams       
0            140
1             73
2             88
3             73
4             99
5             80
6             85
7             89
8            125
9            124
10           125
11           180
12           212
13           293
14           442
15          1040
16          2380

In [17]:
# Yikes! 140 entities in the test set were so hard to find, they
# were extracted by 0 teams.
# Let's go back and look at some of those 0-team entities in context:
difficult_recall["context"] = difficult_recall["span"].apply(lambda t: t.context())
pd.set_option('max_colwidth', 100)
difficult_recall.head(20)

fold  doc_offset                                   span ent_type  gold  \
0   test         163                [587, 596): 'Newsnight'     MISC  True   
1   test         152                     [1487, 1489): 'Po'      LOC  True   
2   test         170                [461, 470): 'Myos Yang'     MISC  True   
3   test         176                 [57, 67): 'K.T. Arasu'      LOC  True   
4   test         199         [108, 124): 'Scottish premier'     MISC  True   
5   test         199  [27, 52): 'SCOTTISH PREMIER DIVISION'     MISC  True   
6   test          97             [141, 153): 'Bahia Blanca'      LOC  True   
7   test          97                  [133, 140): 'Rosario'      LOC  True   
8   test          97                  [125, 132): 'Quequen'      LOC  True   
9   test          63             [148, 160): 'Conservative'     MISC  True   
10  test           8                [697, 706): 'Yorkshire'      LOC  True   
11  test          79                 [219, 227): 'Daniella'     MISC  True   
12  test         115          [326, 341): 'Outagmie County'      LOC  True   
13  test           8               [239, 249): 'McDonald's'      ORG  True   
14  test         222                   [108, 114): 'League'      ORG  True   
15  test          82                [1843, 1850): 'Yakomas'     MISC  True   
16  test         155                  [776, 783): 'Antwerp'      ORG  True   
17  test          38              [1117, 1126): 'Salamanca'      ORG  True   
18  test         186           [2292, 2304): 'Nation's Cup'     MISC  True   
19  test          98                    [395, 400): 'Kenda'     MISC  True   

    num_teams  \
0           0   
1           0   
2           0   
3           0   
4           0   
5           0   
6           0   
7           0   
8           0   
9           0   
10          0   
11          0   
12          0   
13          0   
14          0   
15          0   
16          0   
17          0   
18          0   
19          0   

                                                                                                context  
0    ... endure, " Marlow told BBC television's [Newsnight] programme on Thursday. "\nIt is not sust...  
1            ...ee-day " independence " march along the [Po] River in September, culminating in a de...  
2   ...Sangthai Glory 22/11/96 3,000 Singapore\n[Myos Yang] 5 22/11/96 4,000 Indonesia\nBudisuryana ...  
3   ... Barrick said to continue Busang talks.\n[K.T. Arasu]\nJAKARTA 1996-12-06\nCanada's Bre-X Min...  
4   ...W 1996-12-07\nLeading goalscorers in the\n[Scottish premier] division after Saturday's matche...  
5                              [SCOTTISH PREMIER DIVISION] SCORERS.\nGLASGOW 1996-12-07\nLeading goa...  
6   ...price fix:\nBuenos Aires Quequen Rosario [Bahia Blanca]\nOats unq unq unq unq\nWheat 121 130 ...  
7    ...ember 5 price fix:\nBuenos Aires Quequen [Rosario] Bahia Blanca\nOats unq unq unq unq\nWheat...  
8    ...\nAvg December 5 price fix:\nBuenos Aires [Quequen] Rosario Bahia Blanca\nOats unq unq unq u...  
9   ...ajor's office said on Friday that rebel [Conservative] MP Sir John Gorst had not " resigned t...  
10   ...e in the side.\nBowyer, who moved to the [Yorkshire] club in August for 3.5 million pounds (...  
11     ... centre of the intense tropical cyclone [Daniella] was 570 km (310 miles) north by northwe...  
12  ..., to a 10-year prison term on Thursday, [Outagmie County] Circuit Court Judge Dennis Luebke s...  
13  ...taurant staff during a disturbance at a [McDonald's] fast-food restaurant.\nBowyer, 19, who w...  
14     ...1996-12-07\nStandings of National Hockey\n[League] teams after games played on Friday (tab...  
15     ...rench-owned hotel was slightly damaged.\n[Yakomas] are hounded in stronghold districts of ...  
16      ...kish man smuggled heroin from Turkey to [Antwerp] from where it was taken to Spain, Franc...  
17   ...aulo Bento (Oviedo, Spain), Jose Taira ([Salamanca], Spain):\nForwards- Antonio Folha (Port

**Some of these entities are "difficult" because the test set contains incorrect labels.**

For reference, there's a copy of the CoNLL labeling rules in this repository at
[resources/conll_03/ner/annotation.txt](../resources/conll_03/ner/annotation.txt)

There are 4 incorrect labels in this first set of 20:
* `[3289, 3299): 'Full Light'` should be "Zywiec Full Light"
* `[11, 19): 'Honda RV'` should be tagged `ORG`
* `[1525, 1541): 'Consumer Project'` should be "Consumer Project on Technology" and should be tagged `ORG`
* `[244, 255): 'McDonald 's'` should be tagged `MISC` (because it's an "adjective ... derived from a word which is ... organisation")

In [18]:
# Let's look at the entities that are difficult from the perspective of 
# precision: that is, in many models' results, but not in the gold standard.
difficult_precision = counts[counts["gold"] == False].sort_values("num_teams", ascending=False).reset_index(drop=True)

# Again, we can add some context to these spans:
difficult_precision["context"] = difficult_precision["span"].apply(lambda t: t.context())
difficult_precision.head(20)

fold  doc_offset                                               span  \
0   test           0                                  [19, 24): 'JAPAN'   
1   test         149                           [1504, 1512): 'Consumer'   
2   test         149  [1487, 1520): 'Washington-based Consumer Project'   
3   test         149                           [1100, 1108): 'Internet'   
4   test         149                              [902, 909): 'Western'   
5   test         149                              [902, 909): 'Western'   
6   test         149                             [155, 163): 'Internet'   
7   test         149                                  [65, 70): 'Kaban'   
8   test         149                                   [60, 64): 'Elif'   
9   test         149                             [60, 70): 'Elif Kaban'   
10  test         148                              [715, 722): 'Stephen'   
11  test         148                                 [700, 704): 'Cito'   
12  test         148                                 [700, 704): 'Cito'   
13  test         148                                 [600, 604): 'Cito'   
14  test         148                                 [600, 604): 'Cito'   
15  test         148               [562, 584): 'Activities Association'   
16  test         148                          [573, 584): 'Association'   
17  test         148                           [551, 561): 'New Mexico'   
18  test         148        [555, 584): 'Mexico Activities Association'   
19  test         148                                  [551, 554): 'New'   

   ent_type   gold  num_teams  \
0      MISC  False         16   
1      MISC  False         16   
2      MISC  False         16   
3       ORG  False         16   
4       ORG  False         16   
5      MISC  False         16   
6       ORG  False         16   
7       PER  False         16   
8       ORG  False         16   
9       ORG  False         16   
10      PER  False         16   
11      ORG  False         16   
12      LOC  False         16   
13      ORG  False         16   
14      LOC  False         16   
15      ORG  False         16   
16      ORG  False         16   
17      LOC  False         16   
18      ORG  False         16   
19      LOC  False         16   

                                                                                                context  
0                                                    [JAPAN] GET LUCKY WIN, CHINA IN SURPRISE DEFEAT...  
1     ...r lobbyist heading the Washington-based [Consumer] Project on Technology.\n" None of the tr...  
2   ...s Love, a consumer lobbyist heading the [Washington-based Consumer Project] on Technology.\n"...  
3      ...ine works led to a storm of protests by [Internet] companies and critics who say the pacts...  
4      ...ish all the discussions, " a frustrated [Western] delegate said. "\nThey announced they wi...  
5      ...ish all the discussions, " a frustrated [Western] delegate said. "\nThey announced they wi...  
6      ...nce centre built before the dawn of the [Internet], groups of staid officials made a first...  
7      ...abbles overshadow copyright talks.\nElif [Kaban]\nGENEVA 1996-12-06\nIn a gloomy Geneva co...  
8       ...e squabbles overshadow copyright talks.\n[Elif] Kaban\nGENEVA 1996-12-06\nIn a gloomy Gen...  
9   ...e squabbles overshadow copyright talks.\n[Elif Kaban]\nGENEVA 1996-12-06\nIn a gloomy Geneva ...  
10     ...f the school he attends.\nCito's father, [Stephen] Cito, had admitted filing the metal buc...  
11        ...r, regardless of the school he attends.\n[Cito]'s father, Stephen Cito, had admitted fi...  
12        ...r, regardless of the school he attends.\n[Cito]'s father, Stephen Cito, had admitted fi...  
13         ...o Activities Association decided to bar [Cito] from any inter-scholastic competition u...  
14         ...o Activities Association decided to bar [Cito] from any inter-scholastic competition u...  
15  ... forearm.\nOfficials said the N

In [19]:
difficult_precision.loc[[16]]

fold  doc_offset                       span ent_type   gold  num_teams  \
16  test         148  [573, 584): 'Association'      ORG  False         16   

                                                                                                context  
16  ...fficials said the New Mexico Activities [Association] decided to bar Cito from any inter-scho...

**As with the entities in `difficult_recall`, some of these entities in `difficult_precision` are "difficult" because the test set has missing and incorrect labels.**

**13** of these first 20 "incorrect" results are due to missing and incorrect labels:
* `[25, 32): 'BRITISH''` in document 202 should be tagged `MISC`.
* `[1317, 1327): 'Portsmouth'` in document 207 should be tagged `ORG`, not `LOC`.
* `[110, 118): 'Scottish'` in document 199 should be tagged `MISC`
  (or `[28, 53): 'SCOTTISH PREMIER DIVISION'` and 
  `[110, 135): 'Scottish premier division'` should both be tagged `ORG`).
* `[146, 163): 'Santiago Bernabeu'` in document 40 should be tagged `MISC`
  (because the "s" in `[146, 171): 'Santiago Bernabeu stadium'` is not capitalized).
* `[239, 251): 'Philadelphia'` in document 223 should be tagged `ORG`, not `LOC`.
* `[367, 376): 'Karlsruhe'` in document 36 should be tagged `ORG`, not `LOC`.
* `[1003, 1011): 'Congress'` in document 100 should be tagged `ORG`
  (also, `[957, 964): 'Chilean' ==> MISC` should be replaced with 
  `[957, 973): 'Chilean Congress' ==> ORG`).
* `[420, 428): 'Freiburg'` in document 36 should be tagged `ORG`, not `LOC`.
* In document 70, `[186, 211): 'New York Commodities Desk'`, not `[186, 206): 'New York Commodities'`, should be tagged `ORG`.
* `[263, 271): 'St Louis'` in document 223 should be tagged `ORG`, not `LOC`.
* `[788, 795): 'Antwerp'` in document 155 should be tagged `LOC`, not `ORG`.
* In document 112, `[178, 191): 'John Mills Jr'`, not `[178, 188): 'John Mills'`, should be tagged `PER`.
* `[274, 282): 'COLORADO'` in document 223 should be tagged `ORG`.


In [20]:
# Here's the gold standard data for document 155, for example.
# Note line 12.
doc_id = ("test", 155)
gold_standard_spans[doc_id][0:60]

span ent_type
0       [11, 18): 'Belgian'     MISC
1      [64, 72): 'BRUSSELS'      LOC
2       [170, 175): 'Spain'      LOC
3     [230, 237): 'Belgian'     MISC
4     [348, 355): 'Belgian'     MISC
5     [423, 430): 'Antwerp'      ORG
6     [459, 466): 'Belgian'     MISC
7   [537, 546): 'Barcelona'      LOC
8     [605, 612): 'Turkish'     MISC
9     [712, 719): 'Belgium'      LOC
10    [733, 740): 'Turkish'     MISC
11     [766, 772): 'Turkey'      LOC
12    [776, 783): 'Antwerp'      ORG
13      [811, 816): 'Spain'      LOC
14     [818, 824): 'France'      LOC
15    [829, 836): 'Germany'      LOC
16    [903, 910): 'Belgium'      LOC
17   [934, 942): 'European'     MISC
18  [1007, 1014): 'Belgian'     MISC

In [21]:
# The above gold standard spans in context. 
gold_standard_spans[doc_id]["span"].values

<TokenSpanArray>
[    [11, 18): 'Belgian',    [64, 72): 'BRUSSELS',     [170, 175): 'Spain',
   [230, 237): 'Belgian',   [348, 355): 'Belgian',   [423, 430): 'Antwerp',
   [459, 466): 'Belgian', [537, 546): 'Barcelona',   [605, 612): 'Turkish',
   [712, 719): 'Belgium',   [733, 740): 'Turkish',    [766, 772): 'Turkey',
   [776, 783): 'Antwerp',     [811, 816): 'Spain',    [818, 824): 'France',
   [829, 836): 'Germany',   [903, 910): 'Belgium',  [934, 942): 'European',
 [1007, 1014): 'Belgian']
Length: 19, dtype: TokenSpanDtype

In [22]:
# Repeat the steps from the previous cells using the dev set.
# This takes a while.
dev_gold_standard = tp.io.conll.conll_2003_to_dataframes(
    data_set_info["dev"], ["pos", "phrase", "ent"], [False, True, True])
dev_gold_standard = [
    df.drop(columns=["pos", "phrase_iob", "phrase_type"])
    for df in dev_gold_standard
]

dev_gold_standard_spans = {
    ("dev", i): tp.io.conll.iob_to_spans(dev_gold_standard[i]) 
    for i in range(len(dev_gold_standard))}

dev_outputs = { 
    t: tp.io.conll.conll_2003_output_to_dataframes(
        dev_gold_standard, f"{PROJECT_ROOT}/resources/conll_03/ner/results/{t}/eng.testa")
    for t in teams
}  # Type: Dict[str, List[pd.DataFrame]]

dev_output_spans = {
    t: {("dev", i): tp.io.conll.iob_to_spans(dev_outputs[t][i]) 
        for i in range(len(dev_outputs[t]))}
    for t in teams
}  # Type: Dict[str, Dict[Tuple[str, int], pd.DataFrame]]

dev_span_flags = {t: merge_span_sets(t, dev_gold_standard_spans,
                                      dev_output_spans) 
                   for t in teams}  # Type: Dict[Tuple[str, int]: pd.DataFrame]

dev_indicators = {}  # Type: Dict[Tuple[str, int], pd.DataFrame]
for k in dev_gold_standard_spans.keys():
    result = dev_gold_standard_spans[k]
    for t in teams:
        result = result.merge(dev_span_flags[t][k], how="outer")
    #dev_indicators[k] = result.fillna(False)
    dev_indicators[k] = result.infer_objects(copy=False)
    
for df in dev_indicators.values():
    # Convert the teams' indicator columns into a single matrix of 
    # Boolean values, and sum the number of True values in each row.
    vectors = df[df.columns[3:]].values
    nonzero_counts = np.count_nonzero(vectors, axis=1)
    df["num_teams"] = nonzero_counts
    
dev_counts = pd.concat([
    pd.DataFrame({
        "fold": k[0],  # Keys are (collection, offset) tuples
        "doc_offset": k[1],
        "span" : dev_indicators[k]["span"],
        "ent_type": dev_indicators[k]["ent_type"],
        "gold": dev_indicators[k]["gold"],
        "num_teams": dev_indicators[k]["num_teams"]
    })
    for k in dev_indicators.keys()
])

In [23]:
# How many teams found entities from the dev set that are in the gold standard?
(dev_counts[dev_counts["gold"] == True][["num_teams", "span"]]
 .groupby("num_teams").count()
 .rename(columns={"span": "count"}))

count
num_teams       
0             72
1             55
2             43
3             68
4             50
5             61
6             55
7             62
8             71
9             64
10            95
11           129
12           148
13           245
14           363
15           974
16          3387

In [24]:
# How many teams found entities from the dev set that aren't in the gold standard?
(dev_counts[dev_counts["gold"] == False][["num_teams", "span"]]
 .groupby("num_teams").count()
 .rename(columns={"span": "count"}))

count
num_teams       
16          4478

In [25]:
# Merge the results from the two folds
all_counts = pd.concat([counts, dev_counts])
all_counts.head()

fold  doc_offset               span ent_type   gold  num_teams
0  test           0  [19, 24): 'JAPAN'      LOC   True         12
1  test           0  [19, 24): 'JAPAN'     MISC  False         16
2  test           0  [29, 34): 'LUCKY'     MISC  False         16
3  test           0  [29, 34): 'LUCKY'      PER  False         16
4  test           0    [35, 38): 'WIN'      PER  False         16

In [26]:
all_counts.tail()

fold  doc_offset                            span ent_type   gold  num_teams
25  dev         215  [673, 689): 'Atlas Bangladesh'      PER  False         16
26  dev         215        [679, 689): 'Bangladesh'      LOC  False         16
27  dev         215               [849, 852): 'DSE'      ORG   True         16
28  dev         215    [977, 991): 'Dhaka Newsroom'      ORG   True         15
29  dev         215          [983, 991): 'Newsroom'      ORG  False         16

In [27]:
# Write out the results that are in the gold standard but not in
# few teams' outputs.
in_gold_to_write, not_in_gold_to_write = tp.cleaning.analysis.csv_prep(all_counts, "num_teams",gold_col_name='gold',doc_col_name='doc_offset')
in_gold_to_write

num_teams  fold  doc_offset  \
0           0   dev           2   
9           0   dev          15   
6           0   dev          20   
15          0   dev          22   
35          0   dev          22   
..        ...   ...         ...   
41         16  test         230   
45         16  test         230   
46         16  test         230   
47         16  test         230   
49         16  test         230   

                                      corpus_span corpus_ent_type error_type  \
0                               [25, 30): 'ASHES'            MISC              
9           [15, 40): 'AMERICAN FOOTBALL-RANDALL'            MISC              
6                              [90, 96): 'Berlin'            MISC              
15  [213, 244): 'Solidarity Meeting for Sarajevo'            MISC              
35            [826, 847): 'IAAF Grand Prix Final'            MISC              
..                                            ...             ...        ...   
41                      [1031, 1040): 'World Cup'            MISC              
45                        [1108, 1115): 'Germany'             LOC              
46                          [1127, 1132): 'Irish'            MISC              
47                        [1153, 1160): 'England'             LOC              
49                        [1252, 1259): 'England'             LOC              

   correct_span correct_ent_type notes time_started time_stopped time_elapsed  
0                                                                              
9                                                                              
6                                                                              
15                                                                             
35                                                                             
..          ...              ...   ...          ...          ...          ...  
41                                                                             
45                                                                             
46                                                                             
47                                                                             
49                                                                             

[11590 rows x 12 columns]

In [28]:
not_in_gold_to_write

num_teams  fold  doc_offset                             model_span  \
0          16   dev           0             [20, 34): 'LEICESTERSHIRE'   
2          16   dev           0             [20, 34): 'LEICESTERSHIRE'   
4          16   dev           0                       [93, 97): 'West'   
5          16   dev           0                       [93, 97): 'West'   
6          16   dev           0                    [98, 104): 'Indian'   
..        ...   ...         ...                                    ...   
42         16  test         230  [1076, 1097): 'European championship'   
43         16  test         230          [1071, 1084): '1988 European'   
51         16  test         230      [1346, 1363): 'World Cup winning'   
52         16  test         230              [1346, 1355): 'World Cup'   
53         16  test         230                  [1395, 1400): 'Bobby'   

   model_ent_type error_type corpus_span corpus_ent_type correct_span  \
0             LOC                                                       
2             PER                                                       
4             LOC                                                       
5             ORG                                                       
6            MISC                                                       
..            ...        ...         ...             ...          ...   
42           MISC                                                       
43           MISC                                                       
51           MISC                                                       
52           MISC                                                       
53            LOC                                                       

   correct_ent_type notes time_started time_stopped time_elapsed  
0                                                                 
2                                                                 
4                                                                 
5                                                                 
6                                                                 
..              ...   ...          ...          ...          ...  
42                                                                
43                                                                
51                                                                
52                                                                
53                                                                

[10829 rows x 14 columns]

In [29]:
# Write output files.
in_gold_to_write.to_csv("outputs/CoNLL_2_in_gold.csv", index=False)
not_in_gold_to_write.to_csv("outputs/CoNLL_2_not_in_gold.csv", index=False)